File: LeetCode/Sort/Golang/Leetcode775-global-and-local-inversions.go
package sbgo

func isIdealPermutation(nums []int) bool {
    lmin, n := nums[len(nums) - 1], len(nums)
    min := func(i, j int) int {
        if i < j {
            return i
        }
        return j
    }
    for i := n - 2;i > 0;i-- {
        if nums[i - 1] > lmin {
            return false;
        }
        lmin = min(lmin, nums[i])
    }
    return true
}

File: LeetCode/Sort/Rust/Leetcode775-global-and-local-inversions.rs
impl Solution {
    pub fn is_ideal_permutation(nums: Vec<i32>) -> bool {
        let (mut lmin, n) = (nums[nums.len() - 1], nums.len());
        for i in (1..n - 1).rev() {
            if nums[i - 1] > lmin {
                return false;
            }
            lmin = lmin.min(nums[i])
        }
        true
    }
}
